ath79: new factory format for D-Link DIR-842

s-2 opened the pull request:
Currently, factory images for DIR-842 can only be flashed via
the recovery mode, which does not support DHCP and requires using special
browsers that still upload binary form-data as 'application/octet-stream'.

Starting from DIR-842 v3.10EUb02, the stock Web UI will only accept
an updated image format that is encrypted using AES-256.

The AES key and IV are found in the GPL sources officially released by
D-Link (e.g. DIR842C3_GPL312EUb01DLink/tools/encimg/fw_sign_data.c) or
can be extracted from the official images using 'strings /htdocs/cgibin'.

 * Rename 'factory.bin' images to 'factory-recovery.bin'
 * Add 'factory-webflash.bin' (AES-encrypted 'factory-recovery.bin')

Installation:

 * Web UI: http://192.168.0.1, flash factory-webflash.bin
 * Recovery Web UI: Keep reset button pressed during power-on,
   flash factory-recovery.bin via http://192.168.0.1

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 committed:
images: allow seama-seal board signature override

Certain SEAMA devices (e.g. D-Link DIR-842 C3) require different factory
image formats depending on the stock firmware version currently installed
on the device, or whether the bootloader recovery mode is used.

When the board signature is different for the new image format, seama-seal
needs to be called with individual SEAMA_SIGNATURE arguments per image.

Currently, SEAMA_SIGNATURE is a board-static value passed via DEVICE_VARS.

Add optional argument to allow seama-seal to override SEAMA_SIGNATURE.

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 commented:
Is Python generally acceptable for image tools, or should it be rewritten using C and `libcrypto`?
The OpenWRT build prerequisites explicitly state `python3`, and PyCrypto is included in the usual desktop distributions of Python 3, so there are no further dependencies.

Is there any better solution for creating two factory images for the same device that require different values of the signature that is usually passed via `DEVICE_VARS`?

Image Encryption for DIR-859 was removed from the code, as I could not find anyone on the forum to test it on a real device; however the keys worked fine for reproducing the encrypted images officially released by D-Link. Not sure if that would be enough to include it anyways (the current state of flashing via the recovery would not be changed).

adrianschmutzler commented on scripts/alpha_encimg.py:
```diff
@@ -0,0 +1,54 @@
+
+with open(input_file, "wb") as fd:
+    fd.write(encrypted)
+
```
no empty line at EOF

s-2 replied:
That's curious, there were actually two newline characters at the end rather than one, it seems my text editor recognized the Python format and secretly added another one according to PEP8, while only displaying a single blank line. With a single newline, no blank line is shown, same as with git.

ynezz commented on scripts/alpha_encimg.py:
```diff
@@ -0,0 +1,53 @@
+# License GPLv2+: GNU GPL version 2 or later <http://gnu.org/licenses/gpl-2.0.html>
+#
+
+from Crypto.Cipher import AES
```
This import would result in error on our build slaves

```
$ docker run --rm -it openwrtorg/buildslave:latest python3 -c "from Crypto.Cipher import AES"
Traceback (most recent call last):
  File "<string>", line 1, in <module>
ImportError: No module named 'Crypto'
```

s-2 replied:
> This import would result in error on our build slaves

That's good to know, thanks. While the code was tested against both the legacy PyCrypto as well as PyCryptodome, I was wondering whether the buildbots actually use the same "full" python3 installation as if I were to `apt install python3` e.g. on ubuntu.

However `libssl-dev` is listed as a dependency as well, so I presume it should work out of the box when including the AES functions from openssl within C code?

ynezz replied:
Well, you can run the container yourself `docker run --rm -it openwrtorg/buildslave:latest /bin/bash` and check what is there or missing, the Dockerfile is here https://git.openwrt.org/?p=buildbot.git;a=blob;f=docker/buildslave/Dockerfile;h=96919bdbb3dada21eb0a0e8ad564a0f7fb135f50;hb=HEAD

s-2 committed:
ath79: new factory format for D-Link DIR-842, 859

Currently, factory images for DIR-842 and DIR-859 can only be flashed via
the recovery mode, which does not support DHCP and requires using special
browsers that still upload binary form-data as 'application/octet-stream'.

Starting from DIR-842 v3.10EUb02 and DIR-859 v1.07b02, the stock Web UI
will only accept an updated image format that is encrypted using AES-256.

The AES key and IV are found in the GPL sources officially released by
D-Link (e.g. DIR842C3_GPL312EUb01DLink/tools/encimg/fw_sign_data.c) or
can be extracted from the official images using 'strings /htdocs/cgibin'.

 * Rename current 'factory.bin' images to 'recovery.bin'
 * Add new 'factory.bin' (AES-encrypted 'recovery.bin')

Installation:

 * Web UI: http://192.168.0.1, flash factory.bin
 * Recovery Web UI: Keep reset button pressed during power-on,
   flash recovery.bin via http://192.168.0.1

Signed-off-by: Sebastian Schaper <openwrt@sebastianschaper.net>

s-2 commented:
Finally found some time to rewrite this in C.

Rebased to master.

Also re-added DIR-859, since I could test flashing the encrypted images on a real device by now.

adrianschmutzler commented:
> Starting from DIR-842 v3.10EUb02 and DIR-859 v1.07b02, the stock Web UI
> will only accept an updated image format that is encrypted using AES-256.
> 
> The AES key and IV are found in the GPL sources officially released by
> D-Link (e.g. DIR842C3_GPL312EUb01DLink/tools/encimg/fw_sign_data.c) or
> can be extracted from the official images using 'strings /htdocs/cgibin'.
> 
>  * Rename 'factory.bin' images to 'factory-recovery.bin'
>  * Add 'factory-webflash.bin' (AES-encrypted 'factory-recovery.bin')

Is it really necessary to have two images here? Can't we just use the new webflash image for recovery as well?

adrianschmutzler commented on target/linux/ath79/image/generic.mk:
```diff
@@ -735,9 +747,17 @@ define Device/dlink_dir-859-a1
+  DEVICE_PACKAGES :=  kmod-usb2 kmod-ath10k-ct-smallbuffers \
+	ath10k-firmware-qca988x-ct
   SEAMA_SIGNATURE := wrgac37_dlink.2013gui_dir859
+  IMAGE_SIZE := 15872k
```
That reorder and wrapping stuff seems unrelated to this change.

adrianschmutzler commented on tools/firmware-utils/Makefile:
```diff
@@ -25,6 +25,7 @@ define Host/Compile
 	mkdir -p $(HOST_BUILD_DIR)/bin
 	$(call cc,add_header)
 	$(call cc,addpattern)
+	$(call cc,alpha_encimg, -lcrypto)
```
Space before argument is not required.

s-2 commented:
> > Starting from DIR-842 v3.10EUb02 and DIR-859 v1.07b02, the stock Web UI
> > will only accept an updated image format that is encrypted using AES-256.
> > The AES key and IV are found in the GPL sources officially released by
> > D-Link (e.g. DIR842C3_GPL312EUb01DLink/tools/encimg/fw_sign_data.c) or
> > can be extracted from the official images using 'strings /htdocs/cgibin'.
> > 
> > * Rename 'factory.bin' images to 'factory-recovery.bin'
> > * Add 'factory-webflash.bin' (AES-encrypted 'factory-recovery.bin')
> 
> Is it really necessary to have two images here? Can't we just use the new webflash image for recovery as well?

The bootloader indeed does not support the AES-encrypted images, thus D-Link provide separate firmwares for upgrading in two stages, the first one (for recovery and very old versions) being called `MiddleFirmware`.

Curiously, in many countries (where the devices started selling after introduction of the new format), only the encrypted image is officially offered for download from the website, resulting in users being unable to recover their routers (D-Link folks apparently don't seem to know about that, [telling users their devices were defective](http://forums.dlink.com/index.php?topic=75536.0)).

adrianschmutzler commented:
Okay, though I think I'd prefer just factory.bin and recovery.bin ...

s-2 commented:
> Okay, though I think I'd prefer just factory.bin and recovery.bin ...

Also thought about what would be the appropriate naming, since there's actually very few devices in OpenWRT with multiple images (I think I must have encountered that -webflash suffix somewhere around the DD-WRT world).

I wonder if it might be confusing to the user if the new factory.bin images would be incompatible to factory.bin from a previous release (if it were the other way around, i.e. just the newly-added image being named differently, that would be trivial).

Also thought about something like "legacy", to account for the fact that very old versions (e.g. devices that were never updated) will need this image, too.

I'll check other targets for devices with multiple images again, and probably change it to factory and recovery.

adrianschmutzler commented:
FYI: https://patchwork.ozlabs.org/project/openwrt/patch/20200326155654.48317-1-freifunk@adrianschmutzler.de/

I plan to continue working on this soon.

s-2 commented:
> FYI: https://patchwork.ozlabs.org/project/openwrt/patch/20200326155654.48317-1-freifunk@adrianschmutzler.de/
> 
> I plan to continue working on this soon.

So according to the definition, `factory-recovery.bin` would suggest it could be flashed via web ui _or_ recovery, which is no longer the case for current devices.

Changing to `factory.bin` and `recovery`.bin.

blocktrron commented on tools/firmware-utils/src/alpha_encimg.c:
```diff
@@ -0,0 +1,141 @@
+	}
+	else
+	{
+		// force padding 16 bytes if input is already aligned
+		memset(&readbuf, 0, 16);
+		AES_cbc_encrypt(readbuf, encbuf, 16, &enc_key, aes_iv, AES_ENCRYPT);
+		fwrite(&encbuf, 1, 16, output_file);
```
Validate the input file has the minimum required length for performing these operations.

s-2 replied:
While I agree it would be quite pointless for the user to supply a 0-bytes file (which should result in a 16 bytes file of encrypted padding), this could most probably happen as a subsequent error, and as such should indeed terminate the build.
Or could there even be further issues (e.g. memory leak) with this that I overlooked?

s-2 commented on tools/firmware-utils/src/alpha_encimg.c:
```diff
@@ -0,0 +1,141 @@
+	unsigned char *key = argv[3];
+	unsigned char *iv = argv[4];
+
+	unsigned char readbuf[4096];
+	unsigned char encbuf[4096];
```
Amazing how many occurrences there were for this value, thanks.

s-2 commented:
improved code comments, added SPDX-License-Identifier according to #3449.
changed order of factory / recovery Makefile rules (apparently broke build for multiple images in latest master).
rebased to master.

s-2 commented:
rebased to resolve merge conflict (alphabetic sorting of `image-commands.mk`)

s-2 commented:
Rebased to resolve merge conflict in `tools/firmware-utils/Makefile` due to 1a55a2b1121106063fa1896d35340e638b7e2e3d

Added `-Wall` to `alpha_encimg`...

s-2 commented:
... and fixed warnings (pointer signedness for arguments; parentheses around `fread()` in while-loop).

s-2 commented:
updated copyright note / SPDX as per discussion in PR #4174.

s-2 commented:
rebased to resolve merge conflict (`check-size` after `append-metadata`; cce8d16bf18909e1634c0aeac05e40a7be96a519)

s-2 commented:
rebased to master, introduced code style changes as suggested by running `scripts/checkpatch.pl`

s-2 commented:
c.f. https://github.com/openwrt/openwrt/pull/4174#issuecomment-1012926578
